src/cargo/ops/cargo_package.rs

```diff
@@ -40,23 +43,88 @@ pub struct PackageOpts<'cfg> {
 const ORIGINAL_MANIFEST_FILE: &str = "Cargo.toml.orig";
 const VCS_INFO_FILE: &str = ".cargo_vcs_info.json";
 
+/// A file to archive, implements `Ord` because its used inside of `ArchivedFiles`
+/// this way the inner `BTreeSet` keeps them always ordered, which is needed. Also
+/// implements `Eq` because its needed that there are not two archived files that
+/// point to the same relative path
+#[derive(Clone)]
 struct ArchiveFile {
     /// The relative path in the archive (not including the top-level package
     /// name directory).
     rel_path: PathBuf,
+
     /// String variant of `rel_path`, for convenience.
     rel_str: String,
+
     /// The contents to add to the archive.
     contents: FileContents,
 }
 
+impl PartialEq for ArchiveFile {
+    fn eq(&self, other: &Self) -> bool {
+        self.rel_path.eq(&other.rel_path)
+    }
+}
+
+impl Eq for ArchiveFile {}
+
+impl PartialOrd for ArchiveFile {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.rel_path.cmp(&other.rel_path))
+    }
+}
+
+impl Ord for ArchiveFile {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
+
+/// The sequence of files to archive, always ordered and doesn't contain duplicates
+/// because of the inner `BTreeSet`
+struct ArchivedFiles(BTreeSet<ArchiveFile>);
```

Comment:
I refactored `ArchiveFiles` and the old `Vec<ArchiveFile>` to improve performance, before half the inserts were checked they didn't have duplicates, wich was `O(n)` and at the end they were sorted by path, now they are at all time sorted, and insertions are much cheaper, maybe for very big projects it may have a little impact

Reply:
I haven't yet got a time to review the whole patch thoroughly, but this refactor makes me feel uncertain. It indeed has a better time complexity, but also adds a relative amount of code. Not to mention deriving a few more `Clone`s.

I'd suggest focusing on what we try to resolve in this pull request. We could always go back here to do benchmarks and improve this code path.